lib/web/mage/validation.js

```diff
@@ -980,9 +980,9 @@
         ],
         'validate-code': [
             function (v) {
-                return $.mage.isEmptyNoTrim(v) || /^[a-z]+[a-z0-9_]+$/.test(v);
+                return $.mage.isEmptyNoTrim(v) || /^[a-zA-Z]+[a-zA-Z0-9_]+$/.test(v);
```

Comment:
What about adding the `{0,30}` constraint also in the JavaScript regex? Just to be consistent with PHP validation.
Speaking of which, is it correct that the range starts at `0`? I think an empty attribute code shouldn't be allowed.